java-tools/OsmAndServerUtilities/src/main/java/net/osmand/server/osmgpx/DownloadOsmGPX.java

```diff
@@ -216,12 +214,197 @@ public static void main(String[] args) throws Exception {
+		PreparedStatement updateStmt = dbConn.prepareStatement(
+				"UPDATE " + GPX_METADATA_TABLE_NAME + " SET activity = ? WHERE id = ?"
+		);
+
+		int batchSize = 0;
+		final int BATCH_LIMIT = 1000;
+		int processedCount = 0;
+
+		try (Statement selectStmt = dbConn.createStatement();
+		     ResultSet rs = selectStmt.executeQuery(
+				     "SELECT id, name, description, tags FROM " + GPX_METADATA_TABLE_NAME + " WHERE activity IS NULL LIMIT " + BATCH_LIMIT
+		     )) {
+
+			while (rs.next()) {
+				long id = rs.getLong("id");
+				String name = rs.getString("name");
+				String description = rs.getString("description");
+				Array tagsArray = rs.getArray("tags");
+				List<String> tags = new ArrayList<>();
+
+				if (tagsArray != null) {
+					try (ResultSet tagRs = tagsArray.getResultSet()) {
+						while (tagRs.next()) {
+							tags.add(tagRs.getString(2).toLowerCase());
+						}
+					}
+				}
+
+				String activity = analyzeActivity(name, description, tags, rootPath);
+				if (activity == null) {
+					try (Statement dataStmt = dbConn.createStatement();
+					     ResultSet rf = dataStmt.executeQuery(
+							     "SELECT data FROM " + GPX_FILES_TABLE_NAME + " WHERE id = " + id
+					     )) {
+						if (rf.next()) {
+							byte[] bytes = rf.getBytes("data");
+							activity = analyzeActivityFromGpx(bytes);
+						}
+					}
+				}
+
+				updateStmt.setString(1, activity);
+				updateStmt.setLong(2, id);
+				updateStmt.addBatch();
+
+				batchSize++;
+				processedCount++;
+
+				if (batchSize >= BATCH_LIMIT) {
+					updateStmt.executeBatch();
+					dbConn.commit();
+					batchSize = 0;
+					LOG.info("Processed " + processedCount + " records so far...");
+				}
+			}
+
+			if (batchSize > 0) {
+				updateStmt.executeBatch();
+				dbConn.commit();
+			}
+		}
+		updateStmt.close();
+		LOG.info("Finished populating the 'activity' column. Total records processed: " + processedCount);
+	}
+
+	private String analyzeActivity(String name, String desc, List<String> tags, String rootPath) {
+		Map<String, List<String>> activitiesMap = createActivitiesMap(rootPath);
+		if (activitiesMap.isEmpty()) {
+			return null;
+		}
+
+		Map<String, String> tagMap = new LinkedHashMap<>();
+
+		activitiesMap.forEach((activityId, tagList) ->
+				tagList.stream()
+						.sorted((tag1, tag2) -> Integer.compare(tag2.length(), tag1.length()))
+						.forEach(tag -> tagMap.put(tag, activityId))
+		);
+
+		for (Map.Entry<String, String> entry : tagMap.entrySet()) {
+			String tag = entry.getKey();
+			String activityId = entry.getValue();
+			if (tags.contains(tag) || name.toLowerCase().contains(tag) || desc.toLowerCase().contains(tag)) {
+				return activityId;
+			}
+		}
+		return null;
+	}
+
+	private static Map<String, List<String>> createActivitiesMap(String rootPath) {
+		File activityFile = new File(rootPath, "resources/poi/activities.json");
+		if (!activityFile.exists()) {
+			return Collections.emptyMap();
+		}
```

Comment:
Please avoid calling this method for each track. Just do it one-off at the start of the utility.